flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/SocketTextStreamFunction.java

```diff
@@ -96,19 +96,21 @@ public void run(SourceContext<String> ctx) throws Exception {
 				socket.connect(new InetSocketAddress(hostname, port), CONNECTION_TIMEOUT_TIME);
 				BufferedReader reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
 
-				int data;
-				while (isRunning && (data = reader.read()) != -1) {
-					// check if the string is complete
-					if (data != delimiter) {
-						buffer.append((char) data);
-					}
-					else {
+				char[] cbuf = new char[8192];
+				int byteRead;
+				while (isRunning && (byteRead = reader.read(cbuf)) != -1) {
+					buffer.append(cbuf, 0, byteRead);
+					int delimPos;
+					while (buffer.length() >= delimiter.length() && (delimPos = buffer.indexOf(delimiter)) != -1) {
+						String token = buffer.substring(0, delimPos);
 						// truncate trailing carriage return
-						if (delimiter == '\n' && buffer.length() > 0 && buffer.charAt(buffer.length() - 1) == '\r') {
-							buffer.setLength(buffer.length() - 1);
+						if (delimiter.equals("\n") && token.endsWith("\r")) {
+							token = token.substring(0, token.length() - 1);
+						}
+						if (!token.isEmpty()) {
+							ctx.collect(token);
```

Comment:
This changes existing behavior. Empty strings are filtered out, which wasn't done before.